### ISSUE TITLE
Keyboard navigation does not work properly on Select with a grouped list

### ISSUE THREAD

**Issue by anatoliy-t7** [chakra-ui/zag#2543]
# 🐛 Bug report
I use Select component with svelte (https://zagjs.com/components/svelte/select)

And I use a grouped list in it, but keyboard navigation does not work properly. It jumps from one an iten of one group to other an item of other group

```
<script lang="ts">
  // https://zagjs.com/components/svelte/select
  import * as select from '@zag-js/select';
  import { portal, useMachine, normalizeProps } from '@zag-js/svelte';
  import { fade } from 'svelte/transition';
  import type { HTMLSelectAttributes } from 'svelte/elements';
  import { onMount } from 'svelte';

  type Placement =
    | 'top'
    | 'top-start'
    | 'top-end'
    | 'right'
    | 'right-start'
    | 'right-end'
    | 'bottom'
    | 'bottom-start'
    | 'bottom-end'
    | 'left'
    | 'left-start'
    | 'left-end';

  type Option = {
    [key: string]: string | number;
  };

  type Group = {
    label: string;
    value: string | number;
    group: string;
  };

  type Props = Omit<HTMLSelectAttributes, 'size'> & {
    value: string | number | null | undefined;
    options: Option[] | string[];
    label: string;
    name?: string;
    class?: string;
    labelClass?: string;
    size?: 'sm' | 'md';
    disabled?: boolean;
    loading?: boolean;
    placeholder?: string;
    translateField?: 'value' | 'label' | 'name';
    labelField?: string;
    valueField?: string;
    userPortal?: boolean;
    position?: Placement;
    required?: boolean;
    grouped?: boolean;
    hideLabelVisually?: boolean;
    onchange?(value: string | number): void;
  };

  let {
    value = $bindable(),
    options,
    label,
    name = 'select',
    class: className = '',
    labelClass = '',
    size = 'md',
    disabled = false,
    loading = false,
    placeholder,
    translateField,
    labelField = 'label',
    valueField = 'value',
    userPortal = true,
    position = 'bottom-end',
    required = false,
    grouped = false,
    hideLabelVisually = false,
    onchange,
    ...others
  }: Props = $props();

  const id = $props.id();

  function items(): { value: string; label: string }[] {
    const array = Object.values(options);

    // Check if first item is a string
    const isStringArray = typeof array[0] === 'string';

    return isStringArray
      ? array.map((type) => ({
          value: type,
          label: type.charAt(0).toUpperCase() + type.slice(1).toLowerCase()
        }))
      : array.map((type) => ({
          value: String(type[valueField]),
          label: type[labelField]
        }));
  }

   function isEmpty<T>(v: T): boolean {
    if (v === null || typeof v === 'undefined' || v === undefined) return true;
    if (typeof v === 'string' && v === '') return true;
    if (Array.isArray(v) && v.length === 0) return true;
    if (typeof v === 'object' && v !== null && Object.keys(v).length === 0) return true;
    return false;
    }

  const groupedOptions = $derived.by(() => {
    return options?.reduce(
      (groups, type) => {
        if (typeof type === 'string') {
          const groupKey = 'default';
          if (!groups[groupKey]) {
            groups[groupKey] = [];
          }
          groups[groupKey].push({
            value: type,
            label: type.charAt(0).toUpperCase() + type.slice(1).toLowerCase()
          });
        } else {
          const groupKey = (type as Group).group || 'default';
          if (!groups[groupKey]) {
            groups[groupKey] = [];
          }
          groups[groupKey].push({
            value: String(type[valueField]),
            label: type[labelField] as string
          });
        }
        return groups;
      },
      {} as Record<string, { value: string; label: string }[]>
    );
  });

  let collection = select.collection({
    items: items(),
    itemToValue: (item: { value: string; label: string }) => item.value,
    itemToString: (item: { value: string; label: string }) => item.label
  });

  let listElement = $state<HTMLUListElement>();

  const service = useMachine(select.machine, {
    id,
    defaultValue: !isEmpty(value) ? [String(value)] : [],
    name,
    collection,
    required,
    positioning: { placement: position },
    disabled: disabled || loading,
    loopFocus: true,
    onValueChange(details) {
      value = details.value[0];

      onchange?.(value);
    },

    scrollToIndexFn(details) {
      const el = listElement?.querySelectorAll('li').item(details.index);

      el?.scrollIntoView({
        block: 'nearest'
      });
    }
  });

  const api = $derived(select.connect(service, normalizeProps));

  let selected = $derived(items().find((option) => String(option.value) === String(api.value[0])));

  let buttonElement = $state(undefined) as HTMLButtonElement | undefined;

  let width = $derived.by(() => {
    if (!buttonElement) return 240;
    if (options.length <= 5) {
      return buttonElement ? buttonElement?.clientWidth : 300;
    }
    return (buttonElement ? buttonElement?.clientWidth : 300) - 10;
  });

  onMount(() => {
    buttonElement?.removeAttribute('aria-labelledby');
  });
</script>

<div class={[className]}>
  <div {...api.getRootProps()}>
    <div {...api.getControlProps()} class="relative z-auto w-full">
      <button
        bind:this={buttonElement}
        {...api.getTriggerProps()}
        aria-expanded={api.open}
        aria-label={label}
        type="button"
        class="input text-left flex justify-between items-center gap-2 whitespace-nowrap w-full overflow-hidden disabled:bg-gray-100"
        class:button-sm={size === 'sm'}
        class:button-md={size === 'md'}
      >
        <div>
          {#if selected}
            {#if translateField}
              {translateField === 'name' ? t(selected.label) : t(selected[translateField])}
            {:else}
              {selected?.label}
            {/if}
          {:else if !label}
            <span class="text-light">{placeholder}</span>
          {/if}
        </div>

       
      </button>

      {#if label && !hideLabelVisually}
        <label
          {...api.getLabelProps()}
          class="label {labelClass}"
          class:hasValue={value || value === 0}
 
        >
          {label}
          {#if required}
            <span class="align-text-top">*</span>
          {/if}
        </label>
      {/if}
    </div>
    {#if api.open}
      <div
        use:portal={{ disabled: !userPortal }}
        {...api.getPositionerProps()}
        transition:fade={{ duration: 150 }}
        class=""
      >
        <ul
          bind:this={listElement}
          {...api.getContentProps()}
          class={[
            'w-full h-auto p-1.5 text-sm bg-white border border-gray-300 rounded-lg shadow-xl dark:shadow-none min-w-72 dark:bg-dark-900 dark:border-gray-700 overflow-y-auto overflow-x-hidden scrollbar',
            size === 'sm' && 'text-sm',
            grouped ? 'max-h-[278px]' : 'max-h-[246px]'
          ]}
          style={`min-width: ${width}px;`}
        >
          {#if grouped}
            {#each Object.entries(groupedOptions) as [group, types], index (index)}
              <li class="py-1">
                <h3 class="px-4 py-1 capitalize">{group}</h3>
                <ul {...api.getItemGroupProps({ id: group })}>
                  {#each types as item, i (i)}
                    <li
                      {...api.getItemProps({ item })}
                      class="cursor-pointer select-none py-1 pl-8 pr-4 focus:outline-none flex w-full justify-between gap-2 items-center rounded-md"
                    >
                      <span {...api.getItemTextProps({ item })} class="whitespace-break-spaces">
                      
                          {item.label}
                     
                      </span>
                      <span {...api.getItemIndicatorProps({ item })}>
                       V
                      </span>
                    </li>
                  {/each}
                </ul>
              </li>
            {/each}
          {:else}
            {#each items() as item, index (index)}
              <li
                {...api.getItemProps({ item })}
                class="cursor-pointer select-none py-1 px-4 focus:outline-none flex w-full justify-between gap-2 items-center rounded-md"
              >
                <span {...api.getItemTextProps({ item })} class="whitespace-break-spaces">
                
                    {item.label}
                  
                </span>
                <span {...api.getItemIndicatorProps({ item })}>
                  V
                </span>
              </li>
            {/each}
          {/if}
        </ul>
      </div>
    {/if}
  </div>
 

  <!-- Hidden select for form -->
  <select {...api.getHiddenSelectProps()} {...others} class="hidden">
    {#each items() as option, index (index)}
      <option value={option.value}>
        {option.label}
      </option>
    {/each}
  </select>
</div>

```

**Comment by segunadebayo** [comment]
The select component relied explicitly on the collection.  There's currently a mismatch between the items in the collection and the rendered items.

Here's how to resolve it:

### Use collection's `groupBy`

If you want to group the items, pass the `groupBy` option to the 

```tsx
  const collection =  select.collection<{ value: string; label: string; group?: string }>({
      items: items(),
      itemToValue: (item) => item.value,
      itemToString: (item) => item.label,
      groupBy: (item) => item.group || "default",
    })
```

### Render with the collection groups

Use the `collection.group()` to render the grouped items

```svelte
            {#each collection.group() as [group, types], index (index)}
              <li>
                <h3>{group}</h3>
                <ul {...api.getItemGroupProps({ id: group })}>
                  {#each types as item, i (i)}
                    <li
                      {...api.getItemProps({ item })}>
                      <span {...api.getItemTextProps({ item })}>
                        {item.label}
                      </span>
                      <span {...api.getItemIndicatorProps({ item })}> V </span>
                    </li>
                  {/each}
                </ul>
              </li>
            {/each}
```

Will mention this in the docs shortly.

**Comment by anatoliy-t7** [comment]
Thank you for quick help 🙏
